README.md
# VideoLinkHelper

[![Actions Status](https://github.com/bratsche/video_link_helper/actions/workflows/elixir.yml/badge.svg)](https://github.com/bratsche/video_link_helper/actions?query=workflow%3ACI)

This is a simple library to help deal with embedded video links. It
allows you to extract the ID from a video link, and can use that ID
to generate an iframe to embed the video into a web application.

This library supports Youtube and Vimeo links.

## Installation

This package can be installed by adding `video_link_helper` to your
list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:video_link_helper, "~> 0.2.0"}
  ]
end
```

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at <https://hexdocs.pm/video_link_helper>.

## Usage

Extract an identifier from a Youtube or Vimeo URL:

```elixir
VideoLinkHelper.extract_id("https://www.youtube.com/watch?v=dQw4w9WgXcQ")
```

This will return a tuple containing the video type and the ID of the video,
for example:

```elixir
{:youtube, "dQw4w9WgXcQ"}
```

A video can be embedded on a page in one of two ways:

 - Using the function component:
```elixir
import VideoLinkHelper.VideoComponent

~H"""
<.video id="dQw4w9WgXcQ" video_type="youtube" width="560" height="315" />
"""
```

 - Using the raw HTML function:
 ```elixir
raw(VideoLinkHelper.render_embed("dQw4w9WgXcQ", width: 560, height: 315))
 ```

## License

This is licensed under the [MIT license](LICENSE.md).


CHANGELOG.md
# Changelog

## 0.2.0 (2022-10-02)

### Enhancements

  - New VideoComponent for use from heex templates.
  - New VideoLinkHelper.render_embed/2 API.

### Deprecations

  - VideoLinkHelper.create_youtube_embed/3 - use VideoLinkHelper.render_embed/2
  - VideoLinkHelper.create_vimeo_embed/3 - use VideoLinkHelper.render_embed/2

### Bug Fixes

  - Fixed "too many redirects" issue in Safari when on http.

## 0.1.0

Initial release.

test/video_component_test.exs
defmodule VideoComponentTest do
  use ExUnit.Case
  import Phoenix.LiveViewTest

  doctest VideoLinkHelper.VideoComponent

  describe "VideoComponent" do
    test "youtube video" do
      assert render_component(&VideoLinkHelper.VideoComponent.video/1,
               id: "dQw4w9WgXcQ",
               width: 560,
               height: 315
             ) ==
               "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/dQw4w9WgXcQ\" style=\"border: none;\" allowfullscreen></iframe>"
    end

    test "vimeo video" do
      assert render_component(&VideoLinkHelper.VideoComponent.video/1,
               id: "7819433",
               video_type: "vimeo",
               width: 560,
               height: 315
             ) ==
               "<iframe width=\"560\" height=\"315\" src=\"https://player.vimeo.com/video/7819433\" style=\"border: none;\" allowfullscreen></iframe>"
    end
  end
end


test/video_link_helper_test.exs
defmodule VideoLinkHelperTest do
  use ExUnit.Case
  doctest VideoLinkHelper

  describe "video links" do
    @youtube_links [
      "http://www.youtube.com/v/0zM3nApSvMg?fs=1&hl=en_US&rel=0",
      "http://www.youtube.com/embed/0zM3nApSvMg?rel=0",
      "http://www.youtube.com/watch?v=0zM3nApSvMg&feature=feedrec_grec_index",
      "http://www.youtube.com/watch?v=0zM3nApSvMg",
      "http://youtu.be/0zM3nApSvMg",
      "http://www.youtube.com/watch?v=0zM3nApSvMg#t=0m10s",
      "http://youtu.be/0zM3nApSvMg",
      "http://www.youtube.com/embed/0zM3nApSvMg",
      "http://www.youtube.com/v/0zM3nApSvMg",
      "http://www.youtube.com/e/0zM3nApSvMg",
      "http://www.youtube.com/watch?v=0zM3nApSvMg",
      "http://www.youtube.com/?v=0zM3nApSvMg",
      "http://www.youtube.com/watch?feature=player_embedded&v=0zM3nApSvMg",
      "http://www.youtube.com/?feature=player_embedded&v=0zM3nApSvMg"
    ]

    @vimeo_links [
      "https://www.vimeo.com/148751763",
      "https://vimeo.com/148751763",
      "https://player.vimeo.com/video/148751763",
      "https://vimeo.com/148751763?app_id=123456"
    ]

    test "youtube links" do
      ids =
        @youtube_links
        |> Enum.map(fn url -> VideoLinkHelper.extract_id(url) end)

      assert true ==
               Enum.all?(ids, fn x ->
                 x == {:youtube, "0zM3nApSvMg"}
               end)
    end

    test "vimeo links" do
      ids =
        @vimeo_links
        |> Enum.map(fn url -> VideoLinkHelper.extract_id(url) end)

      assert true ==
               Enum.all?(ids, fn x ->
                 x == {:vimeo, "148751763"}
               end)
    end
  end

  describe "rendering embeds" do
    test "default youtube embed" do
      assert VideoLinkHelper.render_embed("dQw4w9WgXcQ") ==
               """
               <iframe width="560" height="315" src="https://www.youtube.com/embed/dQw4w9WgXcQ" style="border: none;" allowfullscreen></iframe>
               """
    end

    test "embed with custom sizes" do
      assert VideoLinkHelper.render_embed("dQw4w9WgXcQ", width: 640, height: 480) ==
               """
               <iframe width="640" height="480" src="https://www.youtube.com/embed/dQw4w9WgXcQ" style="border: none;" allowfullscreen></iframe>
               """
    end

    test "vimeo embed" do
      assert VideoLinkHelper.render_embed("148751763", video_type: :vimeo) ==
               """
               <iframe width="560" height="315" src="https://player.vimeo.com/video/148751763" style="border: none;" allowfullscreen></iframe>
               """
    end
  end
end


mix.exs
defmodule VideoLinkHelper.MixProject do
  use Mix.Project

  @version "0.2.0"
  @description "Elixir utility library for dealing with embedded Youtube and Vimeo videos"

  def project do
    [
      app: :video_link_helper,
      version: @version,
      description: @description,
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      docs: docs(),
      package: package(),
      source_url: "https://github.com/bratsche/video_link_helper"
    ]
  end

  def application do
    []
  end

  defp deps do
    [
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:jason, "~> 1.0"},
      {:phoenix_live_view, "~> 0.18"}
    ]
  end

  defp docs do
    [
      main: "VideoLinkHelper",
      source_ref: "v#{@version}",
      source_url: "https://github.com/bratsche/video_link_helper",
      extras: ["CHANGELOG.md"],
      skip_undefined_reference_warnings_on: ["CHANGELOG.md"]
    ]
  end

  defp package() do
    [
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/bratsche/video_link_helper"
      }
    ]
  end
end


lib/video_link_helper.ex
defmodule VideoLinkHelper do
  @moduledoc """
  A helper for extracting IDs from video (Youtube or Vimeo) URLs and
  rendering embeds for those IDs.

  """

  @youtube_regex ~r{^.*(?:youtu.be\/|v\/|e\/|u\/\w+\/|embed\/|v=)([^#\&\?]*).*}
  @vimeo_regex ~r{(?:http:|https:|)\/\/(?:player.|www.)?vimeo\.com\/(?:video\/|embed\/|watch\?\S*v=|v\/)?(\d*)}

  @doc """
  Extracts an ID from a link and identifies it as either a Youtube or Vimeo link.

  ## Examples
      iex> VideoLinkHelper.extract_id("https://www.youtube.com/watch?v=dQw4w9WgXcQ")
      {:youtube, "dQw4w9WgXcQ"}

  """
  def extract_id(nil), do: nil

  def extract_id(url) do
    cond do
      url =~ @youtube_regex ->
        [_, id] = Regex.run(@youtube_regex, url)
        {:youtube, id}

      url =~ @vimeo_regex ->
        [_, id] = Regex.run(@vimeo_regex, url)
        {:vimeo, id}

      true ->
        nil
    end
  end

  @default_width 560
  @default_height 315
  @default_type :youtube

  @doc """
  Renders HTML for a video embed.

  ## Options

  * `:width` - Width of the video embed. Default is 560.
  * `:height` - Height of the video embed. Default is 315.
  * `:video_type` - The type of video embed to render. Must be
  either `:youtube` or `:vimeo`. If this option is not provided,
  the default is `:youtube`.

  Returns a string.

  """
  @doc since: "0.2.0"
  def render_embed(id, options \\ []) do
    width = Keyword.get(options, :width, @default_width)
    height = Keyword.get(options, :height, @default_height)
    video_type = Keyword.get(options, :video_type, @default_type)

    create_embed(id, height, width, to_string(video_type))
  end

  @doc deprecated: "Use VideoLinkHelper.render_embed/2 instead."
  def create_youtube_embed(id, height, width), do: create_embed(id, height, width, "youtube")

  @doc deprecated: "Use VideoLinkHelper.render_embed/2 instead."
  def create_vimeo_embed(id, height, width), do: create_embed(id, height, width, "vimeo")

  # -- private functions

  defp create_embed(id, height, width, "youtube") do
    """
    <iframe width="#{width}" height="#{height}" src="https://www.youtube.com/embed/#{id}" style="border: none;" allowfullscreen></iframe>
    """
  end

  defp create_embed(id, height, width, "vimeo") do
    """
    <iframe width="#{width}" height="#{height}" src="https://player.vimeo.com/video/#{id}" style="border: none;" allowfullscreen></iframe>
    """
  end

  defp create_embed(_id, _height, _width, video_type) do
    raise "Invalid video type: #{inspect(video_type)}"
  end
end
